internal/report/report.go

```diff
@@ -0,0 +1,132 @@
+package report
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"os"
+	"time"
+
+	"github.com/open-policy-agent/opa/plugins/rest"
+	"github.com/open-policy-agent/opa/util"
+	"github.com/open-policy-agent/opa/version"
+)
+
+const (
+	defaultExternalServiceURL = "https://telemetry.openpolicyagent.org"
+)
+
+// Reporter reports the version of the running OPA instance to an external service
+type Reporter struct {
+	body   map[string]string
+	client rest.Client
+}
+
+// DataResponse represents the data returned by the external service
+type DataResponse struct {
+	Latest ReleaseDetails `json:"latest,omitempty"`
+}
+
+// ReleaseDetails holds information about the latest OPA release
+type ReleaseDetails struct {
+	Download      string `json:"download,omitempty"`       // link to download the OPA release
+	ReleaseNotes  string `json:"release_notes,omitempty"`  // link to the OPA release notes
+	LatestRelease string `json:"latest_release,omitempty"` // latest OPA released version
+	OPAUpToDate   bool   `json:"opa_up_to_date,omitempty"` // is running OPA version greater than or equal to the latest released
+}
+
+// New returns an instance of the Reporter
+func New(id string) (*Reporter, error) {
+	r := Reporter{}
+	r.body = map[string]string{
+		"id":      id,
+		"version": version.Version,
+	}
+
+	url := os.Getenv("OPA_TELEMETRY_SERVICE_URL")
+	if url == "" {
+		url = defaultExternalServiceURL
+	}
+
+	restConfig := []byte(fmt.Sprintf(`{
+		"url": %q,
+	}`, url))
+
+	client, err := rest.New(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	r.client = client
+
+	return &r, nil
+}
+
+// SendReport sends the version report to the external service
+func (r *Reporter) SendReport(ctx context.Context) (*DataResponse, error) {
+	rCtx, cancel := context.WithTimeout(ctx, 1*time.Second)
+	defer cancel()
+
+	resp, err := r.client.WithJSON(r.body).Do(rCtx, "POST", "/v1/version")
+	if err != nil {
+		return nil, err
+	}
+
+	defer util.Close(resp)
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		if resp.Body != nil {
+			var result DataResponse
+			err := json.NewDecoder(resp.Body).Decode(&result)
+			if err != nil {
+				return nil, err
+			}
+			return &result, nil
+		}
+		return nil, nil
+	default:
+		return nil, fmt.Errorf("server replied with HTTP %v", resp.StatusCode)
+	}
+}
+
+// IsSet returns true if dr is populated.
+func (dr *DataResponse) IsSet() bool {
+	return dr != nil && dr.Latest.LatestRelease != "" && dr.Latest.Download != "" && dr.Latest.ReleaseNotes != ""
+}
+
+// Slice returns the dr as a slice of key-value string pairs. If dr is nil, this function returns an empty slice.
+func (dr *DataResponse) Slice() [][2]string {
+
+	if !dr.IsSet() {
+		return nil
+	}
+
+	return [][2]string{
+		{"Latest Upstream Version", strings.TrimPrefix(dr.Latest.LatestRelease, "v")},
+		{"Download", dr.Latest.Download},
+		{"Release Notes", dr.Latest.ReleaseNotes},
```

Comment:
Should this include the current version? Or some indication that there is an update?
```
Latest OPA Release
==================

Latest Upstream Version : 0.19.2
Download                : https://openpolicyagent.org/downloads/v0.19.2/opa_darwin_amd64
Release Notes           : https://github.com/open-policy-agent/opa/releases/tag/v0.19.2
```
But its not immediately obvious (without going back up to the start of the REPL) what version you are currently on.

Reply:
Updated:

```
Version Info
============
Current Version         : 0.20.0
Latest Upstream Version : 0.19.2
Download                : https://openpolicyagent.org/downloads/v0.19.2/opa_darwin_amd64
Release Notes           : https://github.com/open-policy-agent/opa/releases/tag/v0.19.2
```